queries/vouch_registry.sql

```diff
@@ -0,0 +1,79 @@
+with
+-- Find permanent version of this query at: https://dune.com/queries/674947
+recognized_bonding_pools (pool, name, initial_funder) as (
+  select *
+  from (
+    values
+        -- This Pool has not yet been funded.
+        -- ('\x8353713b6D2F728Ed763a04B886B16aAD2b16eBD'::bytea, 'Gnosis'),
+        ('\x5d4020b9261F01B6f8a45db929704b0Ad6F5e9E6'::bytea, 'CoW Services', '\x423cec87f19f0778f549846e0801ee267a917935'::bytea)
+    ) as _
+),
+vouches as (
+  select
+    evt_block_time,
+    evt_index,
+    solver,
+    "cowRewardTarget" as reward_target,
+    pool,
+    sender,
+    True as active
+  from cow_protocol."VouchRegister_evt_Vouch"
+    join recognized_bonding_pools
+        on pool = "bondingPool"
+        and sender = initial_funder
+),
+invalidations as (
+  select
+    evt_block_time,
+    evt_index,
+    solver,
+    Null::bytea as reward_target,  -- This is just ot align with vouches to take a union
+    pool,
+    sender,
+    False as active
+  from cow_protocol."VouchRegister_evt_InvalidateVouch"
+    join recognized_bonding_pools
+        on pool = "bondingPool"
+        and sender = initial_funder
+),
+-- At this point we have excluded all arbitrary vouches (i.e. those not from initial funders of recognized pools)
+-- This ranks (solver, pool, sender) by most recent (vouch or invalidation) 
+-- and yields as rank 1, the current "active" status of the triplet.
+ranked_vouches as (
```

Comment:
Yes, it is very involved and there are many things to consider. In the example you gave the query would result in an active solver 0 for bonding pool zero, because the it is the most recent event. If the second event didn't happen then the solver would not be considered an active solver because the only thing that ever happened was a deactivation event. That is to say, I believe the query (in its current state) handles the scenario you described how I would expect it to handle it.

Is there any particular reason you want the first invalidation to be excluded from the "relevant events table"? It would only result the non-active solver not being part of the results...

Reply:
To elaborate on the claims made above, you will see there are two "rankings" (aka sortings) being made here. the first sorts the valid events by activation and deactivation, taking only the most recent "active" records for a given pool which are.

The second then removes the possibility that a solver is bonded by more than one pool by sorting/ranking in a way that takes the earliest. 

Taking the earliest valid active bonded solver eliminates the scenario you described where the solver is bonded by a second pool, but still bonded by the first pool.

Reply:
Ah, you are ordering desc. 

Would the query then work for this
```
Vouch  for solver 0 from bonding pool 0.-> gets rank 2
Vouch  for solver 0 from bonding pool 1. -> this is invalid -> gets rank 1.
Invalidate Vouch  for solver 0 from bonding pool 0. -> gets rank 1
```
current_active_vouches 
=>
Vouch  for solver 0 from bonding pool 1.

valid_vouches
=> Vouch  for solver 0 from bonding pool 1.

although the this vouching was invalid to start with. Probably, I have overseen something else....


Yeah tests would be great. Otherwise, its too much manula checking.. I think here its nice and easy to come up with tests, as we can just replace the tables vouches and invaldiations with actual data tables.

Reply:
In the example you give here first event would result in vouch for pool 0, the second event would result not be picked up by this query because with vouches for the same solver the query only takes the first occurring. Then the third event would result in the second pool being vouched for. So it doesn't matter in which order you invalidate the first pool. The first pool is still the vouched one until the first is invalidated.

Reply:
I will hack together something we can use to test this.

Reply:
Here is the test case for what you described here:

https://dune.com/queries/674947?Scenario_e15077=test

Reply:
Specifically, 

```
test_vouch_events (evt_block_number, evt_index, solver, "cowRewardTarget", "bondingPool", sender) as (
    select * from (
        values 
            (0, 0, '\x50'::bytea, '\xc1'::bytea, '\xb0'::bytea, '\xf1'::bytea),
            (1, 0, '\x50'::bytea, '\xc1'::bytea, '\xb1'::bytea, '\xf2'::bytea)
    ) as _
),
test_invalidation_events (evt_block_number, evt_index, solver, "bondingPool", sender) as (
    select * from (
        values 
            (3, 0, '\x50'::bytea, '\xb0'::bytea, '\xf1'::bytea),
            (3, 0, '\xff'::bytea, '\xff'::bytea, '\xff'::bytea) -- This is here as an invalid event so that we can comment out the above entries
    ) as _
),
test_bonding_pools (pool, name, initial_funder) as (
  select *
  from (
    values
        ('\xb0'::bytea, 'Pool 0', '\xf1'::bytea),
        ('\xb1'::bytea, 'Pool 1', '\xf2'::bytea)
    ) as _
),
```

Describes the scenario:

```
-- Vouch  for solver 0 from bonding pool 0.
-- Vouch  for solver 0 from bonding pool 1.
-- Invalidate Vouch  for solver 0 from bonding pool 0
```

You can even comment out the invalidation to valid vouch change from from one pool to the other.

Reply:
> Then the third event would result in the second pool being vouched for.
In the readme, I specified that all vouches for an active solver will be disregarded. 
>In case of conflicting vouching votes (e.g. two bonding pools vote for the same solver with different cow reward targets), the first vouch will be seen as the valid one and the second vouch will be disregarded.
https://github.com/cowprotocol/solver-vouch-register

So I was under the impression that even after the invalidation of pool 0, the vouch from pool 1 will not be effective, as it was disregarded. A disregarded message would not be suddenly valid, because of another invalidation. 


But I think your overall logic is simpler and works as well. E.g we always just look at the last valid vouch.
 I think I should redefine it in the readme. WDYT?

Reply:
I found a test case where this query doesn't work as expected. Specifically updating cow reward target results in both rows... will fix.